bson/bsoncodec/registry.go

```diff
@@ -115,15 +115,75 @@ func buildDefaultRegistry() *Registry {
 
 // RegisterCodec will register the provided ValueCodec for the provided type.
 func (rb *RegistryBuilder) RegisterCodec(t reflect.Type, codec ValueCodec) *RegistryBuilder {
-	rb.RegisterEncoder(t, codec)
-	rb.RegisterDecoder(t, codec)
+	rb.RegisterTypeEncoder(t, codec)
+	rb.RegisterTypeDecoder(t, codec)
 	return rb
 }
 
-// RegisterEncoder will register the provided ValueEncoder to the provided type.
+// RegisterTypeEncoder will register the provided ValueEncoder for the provided type.
 //
-// The type registered will be used directly, so an encoder can be registered for a type and a
-// different encoder can be registered for a pointer to that type.
+// The type will be used directly, so an encoder can be registered for a type and a different encoder can be registered
+// for a pointer to that type.
+//
+// If the given type is an interface, the encoder will be called when marshalling a type that is that interface. It
+// will not be called when marshalling a non-interface type that implements the interface.
+func (rb *RegistryBuilder) RegisterTypeEncoder(t reflect.Type, enc ValueEncoder) *RegistryBuilder {
+	rb.typeEncoders[t] = enc
+	return rb
+}
+
+// RegisterHookEncoder will register an encoder for the provided interface type t. This encoder will be called when
+// marshalling a type if the type implements t or a pointer to the type implements t. If the provided type is not
+// an interface (i.e. t.Kind() != reflect.Interface), it is ignored and the RegistryBuilder is not modified.
+func (rb *RegistryBuilder) RegisterHookEncoder(t reflect.Type, enc ValueEncoder) *RegistryBuilder {
+	if t.Kind() != reflect.Interface {
```

Comment:
One open question here that I want to make sure we resolve: I don't love this behavior because the user likely made a mistake by using `RegisterHookEncoder` for a non-interface type, but we have no way of reporting this error from the builder. Does it make sense for the driver to panic here?

Reply:
This would happen while the user's setting up, so I think it makes sense.

Reply:
Agreed... this is a configuration error that will prevent the application from working properly.